### ISSUE TITLE
Addressing of Delete Activities

### ISSUE THREAD

**Issue by bmottershead** [mastodon/mastodon#23175]
### Steps to reproduce the problem

On a federated Fediverse server,` a user registers and "follows" a user on a Mastodon server, for example, denizen@social.riverscape.info follows Gargron@mastodon.social.  

### Expected behaviour

Stream of "Creates" and "Updates" related to Gargon@mastodon.social sent to the denizen inbox, or the sharedInbox on social.riverscape.info.

### Actual behaviour

The expected behaviour occurred, but also, thousands of "Delete" messages also are sent, vastly outnumbering the  messages sent pursuant to the "Follow".

### Detailed description

Dozens, occasionally hundreds, of Delete activities **per day** are sent from mastodon.social will be sent to the shared inbox on social.riverscape.info.   If  denizen@social.riverscape.info doesn't have a shared inbox, the Deletes will be sent to their inbox.   Indeed, it seems that the Deletes will be sent to all the inbox of every user on social.riverscape.info which mastodon.social knows about.     This has continue since August 2022.

Each of these Deletes has <some-user>@mastodon.social "deleting" <some-user>@mastodon.social; in other words, themselves.   I don't know if this is actually what is happening, or whether this is how it looks when the Mastodon server or an admin deletes a user for some reason, such as cleaning up spurious "users".  Each Delete is addressed to Public.

All this results from one user on a federated server following one user on one Mastodon server..  How many Deletes would there be if the few riverscape.social users followed users on a lot of mastodon servers?    Would social.riverscape.info be hearing about every darn user deletion in Mastodon-land?   social.riverscape.info would become essentially a Delete activity deleter.

### Specifications

The activitypub spec says that a server may send an activity addressed to "Public" to every sharedInbox which it knows of.    But these Public-addressed Delete messages seem to be sent to every inbox, shared on not, which the server knows about.   This is excessive.

**Comment by ClearlyClaire** [comment]
Account deletions are indeed currently broadcast to the entire known network. This is in order to maximize the chances of account deletions actually being processed across the network, as typically many servers will discover and store information from an account without following it, and Mastodon does not currently keep track of which server has ever seen an account.

You can see #22273 for a proposal to address this issue by maintaining an over-approximation of every server which has ever seen a given actor and sending the `Delete` only to that set. Its performance impact still needs to be properly evaluated.

**Comment by bmottershead** [comment]
What is the negative consequence of just not sending these Deletes?    Or sending them from a "Service" actor which other servers can follow?  Or more likely, not follow, since it is hard to see why servers would be interested in these  Deletes?    

What is a server receiving these Deletes supposed to **do** with them, anyway?  The ActivityPub spec does not place any requirement on servers regarding such Deletes, and doing `**nothing**` not only seems consistent with the ActivityPub spec, but also quite reasonable.    

If, notwithstanding, you are determined that other servers who might have "seen" an actor should be notified if the actor is deleted, I don't see why you have to maintain a list for every known server with which an actor has interacted, as envisioned in #22273.  That doesn't seem very scalable.   Doesn't that list effectively exist already in the form of the addressees of the activities originated by the deleted actor -- the activities in the deleted actor's outbox, in other words?   Does Mastodon not have the "outbox" of an actor?

**Comment by nightpool** [comment]
> What is the negative consequence of just not sending these Deletes? Or sending them from a "Service" actor which other servers can follow? Or more likely, not follow, since it is hard to see why servers would be interested in these Deletes?

The negative consequence of not sending these deletes is that we return to the pre-2022 situation where there are a lot of accounts visible on your local server that *seem* like they're still active, but have actually since been deleted on the remote server, since they may have been e.g. boosted once by one of your followers, or you may have looked them up once because they were @-mentioned in some random post someone made.

> What is a server receiving these Deletes supposed to do with them, anyway?

If you see a Delete for an actor, then you should indicate to any user that views that actor that it's been deleted, as well as treating it like an individual Delete for every activity they've created.

> Doesn't that list effectively exist already in the form of the addressees of the activities originated by the deleted actor -- the activities in the deleted actor's outbox, in other words?

No, because servers may learn about an actor for dozens of other reasons besides being a target of their activities. I've listed a few above, but here are some more:

 - Someone you follow Announced their post, they've had no interaction with you and have no clue you exist.
 - Someone you follow @ mentions them
 - Someone links to them, so you plug their URL into your lookup endpoint so you can see if anyone on your server follows them
 - Someone you followed replied to them once

**Comment by ClearlyClaire** [comment]
> What is the negative consequence of just not sending these Deletes?

ActivityPub is basically a delivery model similar to email except most messages happen to be public. Almost all current implementations will store what they receive and display them to other users depending on context. If a server learns about an account (for instance because someone has `Announce`d their posts), but does not receive a `Delete`, they will keep that copy of the data, and will not delete those publicly-served posts even though the original user has decided to delete their account.

> Or sending them from a "Service" actor

One would have to define that the “Service” actor has authority over the actor and can `Delete` it. There's no such thing defined in ActivityPub.

> What is a server receiving these Deletes supposed to do with them, anyway?

Delete all stored data about this actor.

> If, notwithstanding all that, you are determined that other servers who might have "seen" an actor should be notified if the actor is deleted, I don't see why you have to maintain a list for each actor of every known server with which an actor has interacted, as envisioned in #22273. That doesn't seem very scalable.

#22273 keeps track of an over-approximation of the (possibly shared) inboxes of accounts that have “seen” the actor. This over-approximation uses a bloom filter, a probabilistic structure with a fixed size (here in the PR as it currently stands, this size is of 6kB per local account), which can have false positives but no false negatives. The more external servers have “seen” the actor, the less accurate it will be (that is, the more `Delete`s we will send to servers that have never “seen” it), but this is still orders of magnitude better than sending a `Delete` to the entire known network for every user.

> Doesn't that list effectively exist already in the form of the addressees of the activities originated by the deleted actor -- the activities in the deleted actor's outbox, in other words? Does Mastodon not have the "outbox" of an actor?

Whenever a post is sent to `as:Public`, it is visible to everyone who requests it, meaning that anyone it was not originally explicitly sent to may learn about it and store a copy. So the PR keeps track of who requests the actor (which is a required step to verify any post written by the actor).

**Comment by bmottershead** [comment]
Perhaps until #22273 or something like it is implemented, Mastodon should stop sending every actor deletion to every known actor on the network.    The current setup impacts small servers, those with a few users who are following only a few remote users across a few other servers, with the result that they may end up processing primarily deletions of Mastodon actors which they nothing about and which are completely irrelevant.    As I indicated in the original post, a server with one actor following only one foreign actor (Gargron@mastodon.social) will receive a factor of 3 times as many Delete messages from mastodon.social as actual Gargron-related messages.  Obviously, this ratio will vary depend on how many actors are being followed, how chatty their feeds are, and how many Mastodon servers are involved.   But it is excessive.


All that the ActivityPub spec says about caching of data in both Client/Server and Server/Server interactions is:   "HTTP caching mechanisms [RFC7234] SHOULD be respected when appropriate."     Should it not suffice that if/when a cached foreign actor record needs to be used on a server, it should be refetched from the origin server, if stale?

**Comment by ClearlyClaire** [comment]
> Perhaps until https://github.com/mastodon/mastodon/pull/22273 or something like it is implemented, Mastodon should stop sending every actor deletion to every known actor inbox/sharedInbox in the Fediverse.

This would cause unacceptable failure to delete accounts across the fediverse.

> All that the ActivityPub spec says about caching of data in both Client/Server and Server/Server interactions is: "HTTP caching mechanisms [RFC7234] SHOULD be respected when appropriate." Should it not suffice that if/when a cached foreign actor record needs to be used on a server, it should be refetched from the origin server, if stale?

This is far easier said than done and Mastodon is only one of many implementations that stores data until getting a message that it should be updated or deleted.

**Comment by nightpool** [comment]
> This is relevant not only to S2S implementions; It is also relevant to AP C2S implementantions, since these Delete messages are ending up not only in "shared" inboxes, but also in individual actor inboxes, in the case of actors who don't advertise a sharedInbox.

I'm not sure I understand this complaint. Wouldn't this also be true in your world of e.g. Delete messages sent by people you follow, or Delete messages sent about posts you happen to have previously interacted with? It's up to the recipient's server to handle these messages properly so that they don't impact the C2S client.

The correct way to handle delete messages sent by other servers is as follows:

> The side effect of receiving this is that (assuming the object is owned by the sending actor / server) the server receiving the delete activity SHOULD remove its representation of the object with the same id, and MAY replace that representation with a Tombstone object.

Why would this be inappropriate to receive in an actor's inbox?

**Comment by ClearlyClaire** [comment]
> What other activities addressed to "#Public" is a Mastodon server sending to every known actor inbox or sharedInbox in the known Fediverse?

None, it's only `Delete` activities on actors. We may extend that logic to `Update` activities on actors at some point, but only in select cases, and only after something like #22273 gets merged, so that wouldn't be to the whole known fediverse.

Mastodon and a few other projects ignore (so it doesn't try to fetch the key and verify the activity) `Delete` activities where the `actor` and the `object` are the same and it does not know about the `actor`, if you're wondering how to handle this case.

**Comment by bmottershead** [comment]
>The correct way to handle delete messages sent by other servers is ...

nightpool,  the AP spec doesn't define side-effects for activities arriving in an "inbox". (Other than perhaps "inbox forwarding", though that isn't called a "side-effect" in the spec.)  The sentence you are quoting in the AP spec is from Section 6.4, which is part of "6. Client to Server Interaction", and relates to what the "home server" is expected to do with a Delete from a client arriving in the actor's "outbox".

> It's up to the recipient's server to handle these messages properly so that they don't impact the C2S client.


There is nothing really said in the AP spec about a Server filtering messages in an client inbox, other than:

> The server MAY filter its delivery targets according to implementation-specific rules (for example, spam filtering).


This is a MAY, is "implementation-specific" and the only concrete example is spam filtering, which at the risk of being a bit snarky, I will concede is perhaps relevant to Mastodon flooding the Fediverse with its actor deletions. :-)

**Comment by nightpool** [comment]
> The sentence you are quoting in the AP spec is from Section 6.4, which is part of "6. Client to Server Interaction", and relates to what the "home server" is expected to do with a Delete from a client arriving in the actor's "outbox".

Huh? I took it directly from the server-to-server section of the spec:

<img width="1381" alt="image" src="https://user-images.githubusercontent.com/233815/214883683-458b4727-ec4c-4c18-991f-43706bde95a0.png">

**Comment by bmottershead** [comment]
You are right.  THere is  a similar sentence in Section 6, which confused me.

**Comment by trwnh** [comment]
> One would have to define that the “Service” actor has authority over the actor and can Delete it. There's no such thing defined in ActivityPub.

we could have an extension to define that the actor is `attributedTo` the instance actor (or we could maybe say that the instance actor is the `generator` of each actor? but `attributedTo` seems to be how we signal authority everywhere else...) and then in cases where such an instance actor is set, this signals the actor is managed by some service and that it can be deleted by that service.

assuming you want service-level messages to reach some service-level actor or administrator, more generally, you could say that

- if the actor has a `sharedInbox`, this is a good indicator that the actor is on a shared service and you should not send the Delete to the individual inbox. send it to the `sharedInbox`
- if the actor has no `sharedInbox` but is `attributedTo` some other actor, then send it to that other actor's `inbox`
- otherwise (if the actor has no `sharedInbox` and no `attributedTo`), send it to the original actor's `inbox`

**Comment by bmottershead** [comment]
What I intended with the "service" actor thing was something more like this: The Deletes remain the same as now; that is, the actor and the object of the activity are the deleted actor (that is, it is a self-deletion activity).  But the Delete activities are sent to the inbox "from" a service actor.  Another server or actor can subscribe to the remote using some form of "Follow" to service actor to receive these Deletes, if it chooses to do so.

Since AP activities don't have a "from" propery distinct from the "actor",  the "service" would have to be related to the Deletion activity in some other way.    What I had in mind when I made the suggestion was that the Delete would arrive wrapped in the subscriber inbox with an "Announce" whose "actor" was the service actor.  "Follow" on the service actor would then work as usual.   However,  there are probably a lot of approaches which would work, including the suggestions from **trwnh**

Whichever approach is adopted, servers which don't know about the mechanism, and therefore don't subscribe to the stream, don't  receive the Deletes at all.

What do those servers do instead when or if they eventually try to perform some operation which requires information about an actor, and the information they have about that actor doesn't exist or is stale?    

I am not sure, and it would be good if that were clarified, because even with the stream from Mastodon of actor deletes (and eventually updates), that situation can happen.   Nobody is saying, are they?, that if an actor is deleted, that all the activities related to that actor, every reply to an activity from that actor, that every mention of that actor, that every activity addressed to that actor, etc, also be deleted.   So all of that will still exist even for deleted actors.  And if a client or server can encounter an object that is related to a Deleted actor, and does not know that actor is deleted, can attempt an operation requiring information about the Deleted actor.   

Even with the current approach of flooding the network with the actor's self-deletion, a server or client may not know of the deletion, simply because it was not part of the network at the time the Deletion flooded out.   Perhaps it is new; perhaps it was down at the time; perhaps it existed and was running but it wasn't and perhaps still isn't known to the server which originated the Delete; perhaps there was an error in the originator, the receiver, or the network.  Nothing in AP says clients are only allowed to get activities from one server. Perhaps the server is operating on an activity referring to a deleted actor which the client got out-of-band, from an archive, or copied it from sombody's T-shirt.  

How are clients and servers supposed to deal with activities related to actors which have been deleted> (And since they have to be able to do that, the obvious next question is: whatever that is, why not just let servers  do that all the time, and be able to dispense with the actor-self-deletion stream?)

**Comment by trwnh** [comment]
> Nobody is saying, are they?, that if an actor is deleted, that all the activities related to that actor, every reply to an activity from that actor, that every mention of that actor, that every activity addressed to that actor, etc, also be deleted. So all of that will still exist even for deleted actors.

this is up to implementations on how far they want to go, of course, but it generally makes sense to delete your cache of that actor as well as all objects `attributedTo` that actor (plus activities where `actor` is that actor). i wouldn't go so far as to also delete any of the other stuff -- it's okay to have mentions and replies where the references simply don't resolve anymore. you're free to handle this case however you want. some implementations like misskey *do*, in fact, delete all replies when processing a delete for a status. i think they used to serve a Tombstone previously, but this seems to have changed at some point in the past years (circa 2019-ish?)

> How are clients and servers supposed to deal with activities related to actors which have been deleted> (And since they have to be able to do that, the obvious next question is: whatever that is, why not just let servers do that all the time, letting them dispense with the actor-self-deletion stream?)

again i would recommend simply leaving those broken references. keep only what you need. the "why not" of this bit is entirely an artefact of the widespread behavior where current implementations tend to... just sort of cache things forever? although there are some things like the new "cache retention policy" in mastodon 4.1 that is something of a start in addressing this (though it's nowhere near mature yet). but even so, you need to consider that not everything is public or resolvable. you can't determine that a private or direct activity is "stale" because it was only ever delivered to you once, and you can't fetch it. the analogy here is like if someone deletes their email account, and in doing so, requests that you delete any emails received from them as well. you're not *required* to respect that, but that's the expectation. it's what you `SHOULD` do.
